### PR TITLE
[FIRRTL] Const parsing, emitting, and prim ops

### PR DESCRIPTION
**Pull request by trilorez**
Builds on #5152.

The fir parser now parses const types and parses literals as const.

The fir emitter now emits const types.

Primitive ops now accept mixed const operands and infer const result types
if all input types are const.

### CODE REVIEW

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 1]
```diff
@@ -3045,21 +3052,33 @@ void FEnumCreateOp::print(OpAsmPrinter &printer) {
...
+                              ArrayRef<Type>{getResult().getType()});
 }
 
 ParseResult FEnumCreateOp::parse(OpAsmParser &parser, OperationState &result) {
   auto *context = parser.getContext();
 
   OpAsmParser::UnresolvedOperand input;
   std::string fieldName;
-  Type outputType;
+  mlir::FunctionType functionType;
```
Comment (trilorez) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:3065]: I changed `enumcreate` to use a functional type to allow for const operands for a non-const result. It also matches the other create ops.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 2]
```diff
@@ -880,6 +881,19 @@ ParseResult FIRParser::parseType(FIRRTLType &result, const Twine &message) {
...
+    consumeToken(FIRToken::kw_const);
+    if (failed(parseType(result, message)))
+      return failure();
+
+    auto baseType = result.dyn_cast<FIRRTLBaseType>();
+    if (!baseType)
+      return emitError("only hardware types can be 'const'");
+
+    result = baseType.getConstType(true);
+    return success();
```
Comment (youngar) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:895]: Does this let you parse `const const const const {}`?

File: lib/Dialect/FIRRTL/Transforms/InferResets.cpp [hunk 3]
```diff
@@ -1125,6 +1125,16 @@ LogicalResult InferResetsPass::updateReset(ResetNetwork net, ResetKind kind) {
...
+    } else if (auto constCastOp = dyn_cast<ConstCastOp>(wop);
+               constCastOp &&
+               !areTypesConstCastable(constCastOp.getResult().getType(),
+                                      constCastOp.getInput().getType())) {
+      // Update const cast results
+      auto result = constCastOp.getResult();
+      result.setType(constCastOp.getInput().getType().getConstType(false));
+      for (auto *user : result.getUsers())
+        worklist.insert(user);
+      LLVM_DEBUG(llvm::dbgs() << "- Inferred " << constCastOp << "\n");
```
Comment (youngar) [inline lib/Dialect/FIRRTL/Transforms/InferResets.cpp:1137]: Can you move the const castable check in to the body and add a comment explaining what is going on here?

Comment (trilorez) [reply lib/Dialect/FIRRTL/Transforms/InferResets.cpp]: Updated

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp (hunk 2 above)
Comment (youngar) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp]: Yeah, should probably fix it.

**Review by youngar — APPROVED** [review]

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp (hunk 2 above)
Comment (trilorez) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp]: Updated

**Pull request merged**